Add read/write functions for features2d and normalize naming convention

augustinmanecy opened the pull request:
**Merge with contrib**: https://github.com/opencv/opencv_contrib/pull/3003

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake


```
Xopencv_contrib=xfeatures2d-rw
build_image:Custom=ubuntu:18.04
buildworker:Custom=linux-f1
```

vpisarev commented:
@augustinmanecy, thank you for the patch! We appreciate the new functionality that you've added. However, the pull request does not pass our tests. Can you please retain the original names and also make the added read() & write() methods non-virtual?

augustinmanecy commented:
@vpisarev Ok I'm working on java test corrections, as suggested by @berak in the corresponding [opencv_contrib PR](https://github.com/opencv/opencv_contrib/pull/2997) and will commit a fix soon. 
However I cannot access to details of tests failures (for obscure reasons clicking on [Details](https://build.opencv.org.cn/job/precommit/job/contrib-ubuntu-18.04/495/) finishes on a timeout) Can you precise me which tests I have to check particularly on my side before committing again?

asmorkalov commented:
FASTFeatureDetectorTest and ORBDescriptorExtractorTest in Java read Detector configuration from file. Please check changes there.

augustinmanecy commented:
Hi, I just pushed a new commit which fix Java tests. I also added additionnal check in read methods in order to avoid erasing current value of parameters which are not in the configuration file (if a paramater was missing in the configuration file, it was replaced by 0). 
Normally tests should pass now.

augustinmanecy commented:
Ok, the tests for BRIEF, STAR and SURF (from xfeatures2d) have failed because it was not build with my own opencv_contrib branch: (last commit of augustinmanecy/opencv_contrib referenced as PR #2997)
=> How can I specify that this PR should be build with a specific PR of opencv_contrib?

augustinmanecy commented:
Ok, tests fail for BRIEF, SURF and STAR as it was not built withy corresponding opencv_contrib [opencv_contrib#2997](https://github.com/opencv/opencv_contrib/pull/2997). 
Hw could I specify that this opencv branch have to be build with a specific opencv_contrib PR?

asmorkalov commented:
@augustinmanecy You need to create branch with the same name in contrib/extra and propose PRs there. PRs with the same branch name are tested together.

augustinmanecy commented:
Tests failed for windows 10 but the failling tests are relevant to  opencv_test_videoio / opencv_test_videoio.videoio/videocapture_acceleration.read... I don't think this is due to my changes, what to do now?

alalek commented:
ignore

augustinmanecy commented:
Ok, so does it mean this PR can be merge or is there something else to do?

asmorkalov commented on modules/features2d/misc/java/test/FASTFeatureDetectorTest.java:
```diff
@@ -128,8 +126,8 @@ public void testWrite() {
-//        String truth = "<?xml version=\"1.0\"?>\n<opencv_storage>\n<name>Feature2D.FAST</name>\n<nonmaxSuppression>1</nonmaxSuppression>\n<threshold>10</threshold>\n<type>2</type>\n</opencv_storage>\n";
-        String truth = "<?xml version=\"1.0\"?>\n<opencv_storage>\n</opencv_storage>\n";
+        String truth = "<?xml version=\"1.0\"?>\n<opencv_storage>\n<name>Feature2D.FastFeatureDetector</name>\n<threshold>10</threshold>\n<nonmaxSuppression>1</nonmaxSuppression>\n<type>2</type>\n</opencv_storage>\n";
+//        String truth = "<?xml version=\"1.0\"?>\n<opencv_storage>\n</opencv_storage>\n";
```
Dead lines are not needed.

augustinmanecy replied:
Ok, I'll remove theses dead lines

asmorkalov commented on modules/features2d/misc/java/test/SURFDescriptorExtractorTest.java:
```diff
@@ -29,7 +29,7 @@ protected void setUp() throws Exception {
 
         Class[] cParams = {double.class, int.class, int.class, boolean.class, boolean.class};
-        Object[] oValues = {100, 2, 4, true, false};
+        Object[] oValues = {100, 4, 2, true, false};
```
Why?

augustinmanecy replied:
Ok, it was just to fit default constructor parameters, I keep old values (see final suggestion below)

asmorkalov commented on modules/features2d/misc/java/test/SURFDescriptorExtractorTest.java:
```diff
@@ -89,20 +89,25 @@ public void testEmpty() {
+        //assertEquals(1, extractor.getExtended());
+        //assertEquals(0, extractor.getUpright());
+        //assertEquals(4, extractor.getNOctaves());
+        //assertEquals(2, extractor.getNOctaveLayers());
```
I propose to restore checks.

augustinmanecy replied:
This is additionnal checks I tired to add but I didn't succeed originally. Problem was the following:
 - extractor is of type "Feature2D", so It does not have access to methods of derived class SURF like getExtended()...

Here is my proposed workaround:
```patch
diff --git a/modules/features2d/misc/java/test/SURFDescriptorExtractorTest.java b/modules/features2d/misc/java/test/SURFDescriptorExtractorTest.java
index 2bbaad1e2c..ed287ea074 100644
--- a/modules/features2d/misc/java/test/SURFDescriptorExtractorTest.java
+++ b/modules/features2d/misc/java/test/SURFDescriptorExtractorTest.java
@@ -10,10 +10,11 @@ import org.opencv.test.OpenCVTestCase;
 import org.opencv.test.OpenCVTestRunner;
 import org.opencv.imgproc.Imgproc;
 import org.opencv.features2d.Feature2D;
+import org.opencv.xfeatures2d.SURF;
 
 public class SURFDescriptorExtractorTest extends OpenCVTestCase {
 
-    Feature2D extractor;
+    SURF extractor;
     int matSize;
 
     private Mat getTestImg() {
@@ -29,7 +30,7 @@ public class SURFDescriptorExtractorTest extends OpenCVTestCase {
         super.setUp();
 
         Class[] cParams = {double.class, int.class, int.class, boolean.class, boolean.class};
-        Object[] oValues = {100, 4, 2, true, false};
+        Object[] oValues = {100, 2, 4, true, false};
         extractor = createClassInstance(XFEATURES2D+"SURF", DEFAULT_FACTORY, cParams, oValues);
 
         matSize = 100;
@@ -89,16 +90,16 @@ public class SURFDescriptorExtractorTest extends OpenCVTestCase {
 
     public void testRead() {
         String filename = OpenCVTestRunner.getTempFileName("yml");
-        writeFile(filename, "%YAML:1.0\n---\nname: \"Feature2D.SURF\"\nhessianThreshold: 100.\nextended: 1\nupright: 0\nnOctaves: 4\nnOctaveLayers: 2\n");
+        writeFile(filename, "%YAML:1.0\n---\nname: \"Feature2D.SURF\"\nhessianThreshold: 100.\nextended: 1\nupright: 0\nnOctaves: 2\nnOctaveLayers: 4\n");
 
         extractor.read(filename);
 
         assertEquals(128, extractor.descriptorSize());
-        //assertEquals(100., extractor.getHessianThreshold());
-        //assertEquals(1, extractor.getExtended());
-        //assertEquals(0, extractor.getUpright());
-        //assertEquals(4, extractor.getNOctaves());
-        //assertEquals(2, extractor.getNOctaveLayers());
+        assertEquals(100., extractor.getHessianThreshold());
+        assertEquals(true, extractor.getExtended());
+        assertEquals(false, extractor.getUpright());
+        assertEquals(2, extractor.getNOctaves());
+        assertEquals(4, extractor.getNOctaveLayers());
     }
```
If you agree with this (replacing Feature2D by XXX), I can do the changes for all tests of other Features2d to really check read values with testRead() functions (until now, read values were not checked).

asmorkalov replied:
It's good idea. Please go ahead.

augustinmanecy commented:
I updated tests to check read parameters and added some missing tests, all build successfully for me but the pipeline failed because I didn't created a fork for opencv_extra with branch features2d-rw (I do not understand why previous builds succeed if this fork was mandatory...).
So, is it possible to trigger again the pipeline to check if builds pass now?

asmorkalov commented:
I re-triggered build on pullrequest.opencv.org.

augustinmanecy commented:
Build for iOS failed due to inconsistence between naming conventions in generated codes.
 => It seems that generated code for **armv7-iphoneos** does not follow the same convention of generated code for Java interface. 

The following errors are due to the fact that the parameter structure for **cv::SimpleBlobDetector::Params** is not always defined using the same convention in different files:
* The structure is originally defined as **cv::SimpleBlobDetector::Params** in [features2d.hpp - lines 737-768](https://github.com/opencv/opencv/blob/574a6c83f9724f29c3c99b11ebe54c61e802df6d/modules/features2d/include/opencv2/features2d.hpp#L737-L768) 
* It is defined as **cv::SimpleBlobDetectorParams** in generated _**build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/features2d/SimpleBlobDetectorParams.h:28**_ 
* Where it is attended to be *cv::SimpleBlobDetector_Params* in generated file _**build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/features2d/SimpleBlobDetector.mm:53:44**_.

I would suggest to generate this structure as **cv::SimpleBlobDetector_Params** in _**build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/features2d/SimpleBlobDetectorParams.h:28**_to remains coherent with naming convention chosen in Java interface.

Here is the compilation error obtained on the forge:
```
CompileC /Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc/framework_build/opencv2.build/Release-iphoneos/opencv2.build/Objects-normal/armv7/SimpleBlobDetector.o objc/features2d/SimpleBlobDetector.mm normal armv7 objective-c++ com.apple.compilers.llvm.clang.1_0.compiler
    cd /Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen
    export LANG=en_US.US-ASCII
    export PATH="/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/usr/bin:/Applications/Xcode.app/Contents/Developer/usr/bin:/usr/local/bin:/opt/build/bin:/opt/pythonenv/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin"
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -x objective-c++ -target armv7-apple-ios9.0 -fmessage-length=0 -fdiagnostics-show-note-include-stack -fmacro-backtrace-limit=0 -Wno-trigraphs -fpascal-strings -O1 -Wno-missing-field-initializers -Wno-missing-prototypes -Wno-return-type -Wno-implicit-atomic-properties -Wno-objc-interface-ivars -Wno-arc-repeated-use-of-weak -Wno-non-virtual-dtor -Wno-overloaded-virtual -Wno-exit-time-destructors -Wno-missing-braces -Wparentheses -Wswitch -Wno-unused-function -Wno-unused-label -Wno-unused-parameter -Wno-unused-variable -Wunused-value -Wno-empty-body -Wno-uninitialized -Wno-unknown-pragmas -Wno-shadow -Wno-four-char-constants -Wno-conversion -Wno-constant-conversion -Wno-int-conversion -Wno-bool-conversion -Wno-enum-conversion -Wno-float-conversion -Wno-non-literal-null-conversion -Wno-objc-literal-conversion -Wno-shorten-64-to-32 -Wno-newline-eof -Wno-selector -Wno-strict-selector-match -Wno-undeclared-selector -Wno-deprecated-implementations -Wno-c++11-extensions -DCMAKE_INTDIR=\"Release-iphoneos\" -DOBJC_OLD_DISPATCH_PROTOTYPES=1 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS13.2.sdk -fstrict-aliasing -Wprotocol -Wdeprecated-declarations -Winvalid-offsetof -Wno-sign-conversion -Wno-infinite-recursion -Wno-move -Wno-comma -Wno-block-capture-autoreleasing -Wno-strict-prototypes -Wno-range-loop-analysis -Wno-semicolon-before-method-body -fembed-bitcode -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/lib/Release/include -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/install/include -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/install/include/opencv2 -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/core -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/imgproc -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/ml -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/phase_unwrapping -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/photo -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/plot -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/xphoto -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/dnn -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/features2d -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/imgcodecs -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/text -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/videoio -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/wechat_qrcode -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/calib3d -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/objdetect -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/structured_light -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/video -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/xfeatures2d -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/ximgproc -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/aruco -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/bgsegm -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/bioinspired -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/face -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/tracking -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/img_hash -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc/framework_build/opencv2.build/Release-iphoneos/opencv2.build/DerivedSources-normal/armv7 -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc/framework_build/opencv2.build/Release-iphoneos/opencv2.build/DerivedSources/armv7 -I/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc/framework_build/opencv2.build/Release-iphoneos/opencv2.build/DerivedSources -F/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/lib/Release -fembed-bitcode -fobjc-arc -fobjc-weak -fvisibility=hidden -fPIC -D__OPENCV_BUILD=1 -Wno-incomplete-umbrella -DNDEBUG -std=gnu++11 -MMD -MT dependencies -MF /Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc/framework_build/opencv2.build/Release-iphoneos/opencv2.build/Objects-normal/armv7/SimpleBlobDetector.d --serialize-diagnostics /Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc/framework_build/opencv2.build/Release-iphoneos/opencv2.build/Objects-normal/armv7/SimpleBlobDetector.dia -c /Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/features2d/SimpleBlobDetector.mm -o /Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc/framework_build/opencv2.build/Release-iphoneos/opencv2.build/Objects-normal/armv7/SimpleBlobDetector.o
/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/features2d/SimpleBlobDetector.mm:53:17: error: no member named 'SimpleBlobDetector_Params' in namespace 'cv'
    cv::Ptr<cv::SimpleBlobDetector_Params> retVal = new cv::SimpleBlobDetector_Params(self.nativePtrSimpleBlobDetector->getParams());
            ~~~~^
/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/features2d/SimpleBlobDetector.mm:53:44: error: C++ requires a type specifier for all declarations
    cv::Ptr<cv::SimpleBlobDetector_Params> retVal = new cv::SimpleBlobDetector_Params(self.nativePtrSimpleBlobDetector->getParams());
                                           ^
/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/features2d/SimpleBlobDetector.mm:53:57: error: no type named 'SimpleBlobDetector_Params' in namespace 'cv'; did you mean 'SimpleBlobDetectorParams'?
    cv::Ptr<cv::SimpleBlobDetector_Params> retVal = new cv::SimpleBlobDetector_Params(self.nativePtrSimpleBlobDetector->getParams());
                                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                                        SimpleBlobDetectorParams
In file included from /Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/features2d/SimpleBlobDetector.mm:8:
/Volumes/build-storage/build/precommit-contrib_ios/build/build_ios_contrib/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/features2d/SimpleBlobDetectorParams.h:28:23: note: 'SimpleBlobDetectorParams' declared here
CV_EXPORTS @interface SimpleBlobDetectorParams : NSObject
                      ^
3 errors generated.
```

augustinmanecy commented:
I'm not sure how to manage **"Compare ABI dumps" fails** during build... 

These are due to new pure methods I added for getters/setters in some features2d base class. This getters/setters are implemented in there **XXX_Impl** derived class and are needed for Java tests...

I can make this methods not pure (to not break derived class implementation based on previous version), as it is done for example with [SIFT](https://github.com/opencv/opencv/blob/fae6831402756a10542e693338b1e3dbfd701f6e/modules/features2d/include/opencv2/features2d.hpp#L382-L386).:
```cpp
    /** @brief Set detection threshold.
    @param threshold AGAST detection threshold score.
    */
    CV_WRAP virtual void setThreshold(int threshold) { CV_UNUSED(threshold); return; }
    CV_WRAP virtual int getThreshold() const { return -1; }
```
=>The problem with this solution, is that this is not the initial implementation chosen for base class of other features2d...

Any suggestions?

augustinmanecy commented:
> I'm not sure how to manage **"Compare ABI dumps" fails** during build...
> 
> These are due to new pure methods I added for getters/setters in some features2d base class. This getters/setters are implemented in there **XXX_Impl** derived class and are needed for Java tests...
> 
> I can make this methods not pure (to not break derived class implementation based on previous version), as it is done for example with [BRISK](https://github.com/opencv/opencv/blob/fae6831402756a10542e693338b1e3dbfd701f6e/modules/features2d/include/opencv2/features2d.hpp#L382-L386).:
> 
> ```c++
>     /** @brief Set detection threshold.
>     @param threshold AGAST detection threshold score.
>     */
>     CV_WRAP virtual void setThreshold(int threshold) { CV_UNUSED(threshold); return; }
>     CV_WRAP virtual int getThreshold() const { return -1; }
> ```
> 
> =>The problem with this solution, is that this is not the initial implementation chosen for base class of other features2d...
> 
> Any suggestions?

So, what should I do for this point? Dummy core as done in BRISK, change nothing, or something else?

augustinmanecy commented:
Build for OpenCV CN Ubuntu 18.04 and 20.04 failed in the forge because **DBUILD_opencv_xfeatures2d** is set to OFF:
```
+ cmake -G Ninja -DBUILD_DOCS=ON -DPYTHON_DEFAULT_EXECUTABLE=/usr/bin/python3 -DBUILD_opencv_xfeatures2d=OFF -DOPENCV_DOWNLOAD_PATH=/home/ci/binaries_cache ../opencv
```
But this define need to be set to ON to build features2d Java Tests (see below)...
```
compile:
    [mkdir] Created dir: /home/opencv-cn/slave/workspace/precommit/ubuntu-20.04-arm64/opencv-build/java_test/build/classes
    [javac] Compiling 71 source files to /home/opencv-cn/slave/workspace/precommit/ubuntu-20.04-arm64/opencv-build/java_test/build/classes
    [javac] /home/opencv-cn/slave/workspace/precommit/ubuntu-20.04-arm64/opencv-build/java_test/src/org/opencv/test/features2d/BEBLIDDescriptorExtractorTest.java:5: error: package org.opencv.xfeatures2d does not exist
    [javac] import org.opencv.xfeatures2d.BEBLID;
    [javac]                              ^
    [javac] /home/opencv-cn/slave/workspace/precommit/ubuntu-20.04-arm64/opencv-build/java_test/src/org/opencv/test/features2d/BEBLIDDescriptorExtractorTest.java:9: error: cannot find symbol
    [javac]     BEBLID extractor;
    [javac]     ^
    [javac]   symbol:   class BEBLID
    [javac]   location: class BEBLIDDescriptorExtractorTest
```

augustinmanecy commented:
Hi,
I would like to conclude this work and know if PR can be accepted, I've not hear from you for several weeks.  What should I do to have the PR be accepted?

According to last build,  there are 3 issues I can't solve without some help or guidelines:
   * Build for iOS failed due to inconsistence between naming conventions in generated codes. => See my comment of 20th July
   *  "Compare ABI dumps" fails during build due to adding of some pure method (setter/getters) needed for Javatest I added. See my comment of 23th July
   * Build for OpenCV CN Ubuntu 18.04 and 20.04 failed in the forge because DBUILD_opencv_xfeatures2d is set to OFF => See my comment of 29th July

asmorkalov commented on modules/features2d/misc/java/test/AKAZEDescriptorExtractorTest.java:
```diff
@@ -0,0 +1,93 @@
+        assertEquals(4, extractor.getDescriptorType());
+        assertEquals(2, extractor.getDescriptorChannels());
+        assertEquals(32, extractor.getDescriptorSize());
+        assertEquals((double)((float)0.002), extractor.getThreshold());
```
Looks very suspicious. Please use assertEquals overload for floating point values with epsilon.

asmorkalov commented on modules/features2d/misc/java/test/AKAZEDescriptorExtractorTest.java:
```diff
@@ -0,0 +1,93 @@
+        assertEquals(4, extractor.getDescriptorType());
+        assertEquals(2, extractor.getDescriptorChannels());
+        assertEquals(32, extractor.getDescriptorSize());
+        assertEquals((double)((float)0.002), extractor.getThreshold());
```
The same.

asmorkalov reviewed (commented):
Hello @Palmitoxico Thanks a lot for hard work! The PR looks good to me in general. There are several things that can simplify things and improve the code:
- No need to cast floating point values, you can just add type literal like this (float)0.2 -> 0.2f.
- It make sense to use common base class with "not implemented" exceptions and inherit test classes from it.
- Please do not use (double)(float) casting. JUnit supports assertEqual for floats with epsilon
- write + writeYml test does not give much to test coverage. cv::FileStorage is used internally everywhere, so only one test case is enough.

asmorkalov commented on modules/features2d/include/opencv2/features2d.hpp:
```diff
@@ -705,7 +748,9 @@ class CV_EXPORTS_W SimpleBlobDetector : public Feature2D
       CV_PROP_RW bool filterByColor;
-      CV_PROP_RW uchar blobColor;
+      CV_PROP_RW uchar blobColor; // kept CV_PROP_RW for python export
+      CV_WRAP int get_blobColor() const; // workaround as Java cannot export unsigned types TODO: find a way to not pollute Python API with get_blobColor
+      CV_WRAP void set_blobColor(int _blobColor); // workaround as Java cannot export unsigned types TODO: find a way to not pollute Python API with get_blobColor
```
OpenCV uses signed types for wrappers. Please remove TODO note.

asmorkalov commented on modules/features2d/include/opencv2/features2d.hpp:
```diff
@@ -725,6 +770,10 @@ class CV_EXPORTS_W SimpleBlobDetector : public Feature2D
   CV_WRAP static Ptr<SimpleBlobDetector>
     create(const SimpleBlobDetector::Params &parameters = SimpleBlobDetector::Params());
+
+  CV_WRAP virtual void setParams( SimpleBlobDetector::Params params ) { CV_UNUSED(params); return; }
+  CV_WRAP virtual SimpleBlobDetector::Params getParams() { return SimpleBlobDetector::Params(); }
```
The methods look confusing. Please remove them if they cannot implement parameters management.

Palmitoxico commented:
@asmorkalov I think you mean @augustinmanecy ?

asmorkalov commented:
Thanks a lot! Sure.

augustinmanecy commented:
@asmorkalov 
Do you have a bit of time to take a look to my last changes (commit [8698ece](https://github.com/opencv/opencv/pull/20367/commits/8698eceb4abd55b3765cf93ff91bb5b03a9cd128)) and tell me if it is acceptable to validate PR?

sturkmen72 commented:
I could help to complete this PR if needed.

BTW what is your opinion about using only `Params` to get and set values like [this example usage](https://github.com/opencv/opencv_contrib/blob/4.x/modules/ximgproc/samples/fld_lines.cpp#L77-L80)

augustinmanecy commented:
@sturkmen72, yes with pleasure, I don't really know what is missing now to complete this PR.
Conerning the management of parameters, personally I agree with you and prefer to use a Params structure and a getParams() setParams() method. We could do the same as [blobDetector](https://github.com/augustinmanecy/opencv/blob/ed2a69839293625ecae9f2b9e53eeefc23ceedd6/modules/features2d/src/blobdetector.cpp#L61-L185).
But it will change a lot the API and lead to a lot of *"Compare ABI dumps"* fails if we replace old setters/getters by setParams getParams... If you think this is not a problem, we can go this way!

asmorkalov commented:
@alalek Could you take a look again? Known issues:
- disabled setProperties / getProperties for SimpleBlobDetector in Obj-C, because binding is generated incorrectly.
- SimpleBlobDetector::Params::blobColor is not exported to java as it uses uchar.

Both of the issues coulld be solved by separate PR. Please add supressions to ABI checker. In case if it looks ok, I'll squash commits and merge the PR.

asmorkalov commented:
@alalek Could you take a look on this?

alalek commented on modules/features2d/misc/java/test/FASTFeatureDetectorTest.java:
```diff
@@ -78,20 +77,24 @@ public void testDetectMatListOfKeyPointMat() {
+        String filename = OpenCVTestRunner.getTempFileName("xml");
 
-        writeFile(filename, "%YAML:1.0\n---\nthreshold: 130\nnonmaxSuppression: 1\n");
+        writeFile(filename, "<?xml version=\"1.0\"?>\n<opencv_storage>\n<name>Feature2D.FastFeatureDetector</name>\n<threshold>10</threshold>\n<nonmaxSuppression>1</nonmaxSuppression>\n<type>2</type>\n</opencv_storage>\n");
```
Why is test code changed?
Are old configurations become broken?

asmorkalov replied:
The PR adds adds more parameters to saved configuration. The YAML->XML replacement looks strange, but the line changes anyway. I decided not to rework original solution.

alalek commented on modules/features2d/misc/java/test/ORBDescriptorExtractorTest.java:
```diff
@@ -51,8 +51,8 @@ public void testComputeMatListOfKeyPointMat() {
             {
-                put(0, 0,
-                        6, 74, 6, 129, 2, 130, 56, 0, 44, 132, 66, 165, 172, 6, 3, 72, 102, 61, 171, 214, 0, 144, 65, 232, 4, 32, 138, 131, 4, 21, 37, 217);
+                put(0, 0, 6, 74, 6, 129, 2, 130, 56, 0, 44, 132, 66, 165, 172, 6, 3, 72, 102, 61, 171, 214, 0, 144, 65,
+                        232, 4, 32, 138, 131, 4, 21, 37, 217);
```
unnecessary change.
Also the first two values a data location which is highlighted by formatting.

alalek commented on modules/features2d/include/opencv2/features2d.hpp:
```diff
@@ -256,7 +256,6 @@ class CV_EXPORTS_W AffineFeature : public Feature2D
 typedef AffineFeature AffineFeatureDetector;
 typedef AffineFeature AffineDescriptorExtractor;
 
-
```
unnecessary change

alalek commented on modules/features2d/include/opencv2/features2d.hpp:
```diff
@@ -731,6 +773,10 @@ class CV_EXPORTS_W SimpleBlobDetector : public Feature2D
   CV_WRAP static Ptr<SimpleBlobDetector>
     create(const SimpleBlobDetector::Params &parameters = SimpleBlobDetector::Params());
+
+  CV_WRAP virtual void setParams(const SimpleBlobDetector::Params& params ) = 0;
```
Do we really need `setParams()` if we already have ctor/create?
Especially "virtual".

alalek commented on modules/features2d/src/agast.cpp:
```diff
@@ -7948,6 +7948,27 @@ class AgastFeatureDetector_Impl CV_FINAL : public AgastFeatureDetector
     : threshold(_threshold), nonmaxSuppression(_nonmaxSuppression), type(_type)
     {}
 
+    void read( const FileNode& fn) CV_OVERRIDE
+    {
+      // if node is empty, keep previous value
+      if (!fn["threshold"].empty())
+        fn["threshold"] >> threshold;
+      if (!fn["nonmaxSuppression"].empty())
+        fn["nonmaxSuppression"] >> nonmaxSuppression;
+      if (!fn["type"].empty())
+        fn["type"] >> type;
+    }
+    void write( FileStorage& fs) const CV_OVERRIDE
+    {
+      if(fs.isOpened())
+      {
+        fs << "name" << getDefaultName();
+        fs << "threshold" << threshold;
+        fs << "nonmaxSuppression" << nonmaxSuppression;
+        fs << "type" << type;
+      }
+    }
```
Indentation is 4 spaces in this file.

The same is below.

alalek commented on modules/features2d/misc/objc/gen_dict.json:
```diff
@@ -1,4 +1,10 @@
 {
+    "ManualFuncs" : {
+        "SimpleBlobDetector": {
+            "setParams": { "declaration" : [""], "implementation" : [""] },
+            "getParams": { "declaration" : [""], "implementation" : [""] }
+        }
+    },
```
Does it really work?

asmorkalov replied:
Yes. I checked objc code generation and this block  just disables broken bindings.

asmorkalov replied:
It'll be replaced with appropriate solution with another PR.

asmorkalov commented:
@alalek Can we merge this?

alalek commented:
There is one build warning left on Win64 (MSVS 2015) configuration.

asmorkalov commented:
@alalek ready for merge.

asmorkalov commented:
@alalek may I merge the patch?

seanm commented on modules/features2d/src/blobdetector.cpp:
```diff
@@ -71,6 +71,37 @@ class CV_EXPORTS_W SimpleBlobDetectorImpl : public SimpleBlobDetector
+      if (p.thresholdStep <= 0)
+          CV_Error(Error::StsBadArg, "thresholdStep>0");
+
+      if (p.minThreshold > p.maxThreshold || p.minThreshold <= 0)
```
@augustinmanecy Should that be `p.minThreshold < 0`?

Some of my own internal tests are now failing against OpenCV master, and I think this is why.

asmorkalov replied:
https://github.com/opencv/opencv/pull/23036
